File: Tree/Divide&Conquer/1026. Maximum Difference Between Node and Ancestor/1026. Maximum Difference Between Node and Ancestor.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int rets = 0;
    int maxAncestorDiff(TreeNode* root) {
        if (root == NULL) return rets;
        dfs(root);
        return rets;
    }
    
    pair<int, int> dfs(TreeNode* node) {
        if (node == NULL) return {INT_MIN, INT_MAX};
        
        pair<int, int> left = dfs(node->left);
        pair<int, int> right = dfs(node->right);
        int cur = node->val;
        if (left.first != INT_MIN && left.second != INT_MAX) {
            rets = max(rets, max(abs(cur-left.first), abs(cur-left.second)));
        }
        
        if (right.first != INT_MIN && right.second != INT_MAX) {
            rets = max(rets, max(abs(cur-right.first), abs(cur-right.second)));
        }
        
        return {max(cur, max(left.first, right.first)), min(cur, min(left.second, right.second))};
    }
};

File: Tree/Divide&Conquer/1026. Maximum Difference Between Node and Ancestor/1026. Maximum Difference Between Node and Ancestor.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    int rets = 0;
    public int maxAncestorDiff(TreeNode root) {
        // max min + recursion backtracking
        if (root == null) return rets;
        dfs(root);
        return rets;
    }
    private int[] dfs(TreeNode node) {
        if (node == null) return new int[] {Integer.MIN_VALUE, Integer.MAX_VALUE};
        
        int[] left = dfs(node.left);
        int[] right = dfs(node.right);
        int cur = node.val;
        if (left[0] != Integer.MIN_VALUE && left[1] != Integer.MAX_VALUE) {
            rets = Math.max(rets, Math.max(Math.abs(cur-left[0]), Math.abs(cur-left[1])));
        }
        
        if (right[0] != Integer.MIN_VALUE && right[1] != Integer.MAX_VALUE) {
            rets = Math.max(rets, Math.max(Math.abs(cur-right[0]), Math.abs(cur-right[1])));
        }
        
        return new int[] {Math.max(cur, Math.max(left[0], right[0])), Math.min(cur, Math.min(left[1], right[1]))};
    }
}
